### PR TITLE
Move powershell to `2.0.0-preview3-25426-01` using the .NET CLI `2.0.0-preview2-006502`

### LINKED ISSUE #3649: Avoid using powershell assembly load context with netcoreapp2.0
Issue Summary
----------------
The API `AppDomain.GetAssemblies` is brought back in .NET Core 2.0 which returns the loaded assemblies from the default loader. Therefore it's possible now for powershell to just depend on the default CoreCLR loader without having our own assembly load context getting in the picture. This would greatly simplify the scenario of hosting powershell in applications.

CoreFX Fixes needed
-----------------------
- [x] https://github.com/dotnet/corefx/issues/18989 - `AssemblyLoadContext.GetLoadContext` crashes when pass in a dynamic assembly. **[Verified Fixed]**
- [x] https://github.com/dotnet/corefx/issues/18877 - `AppDomain.GetAssemblies` doesn't return dynamic assemblies that are emitted on the fly. **[Verified Fixed]**
- [x] https://github.com/dotnet/corefx/issues/18791 - For an assembly loaded via 'Assembly.LoadFrom', the assembly it referenced cannot be loaded even if they are in the same folder. **[Verified Fixed]**
- [x] OmniSharp/omnisharp-vscode#1495 - Preview 2 builds are not yet supported by VS Code C#. **[Verified Fixed] -- `v1.11.0-Beta4` VSCode C# extension has the fix and it needs to be installed manually`**
- [x] https://github.com/dotnet/corefx/issues/21095 - [Regression] ResourceManager.GetString(string) throw System.IO.FileNotFoundException. **[Verified Fixed]**

Tasks
-----
- [x] Refactor the code to make PowerShell Core depend on the default CoreCLR loader. Fixed via #3903
- [x] Update the hosting code in [pwrshclrhost.cpp](https://github.com/PowerShell/PowerShell/blob/309bf603f9aff9009dce7e725d42b2d4f99f58c9/src/powershell-native/nativemsh/pwrshplugin/pwrshclrhost.cpp#L62)
- [x] Update the host code in [Shell.cs](https://github.com/PowerShell/psl-omi-provider/blob/master/src/Shell.c#L446) in psl-omi-provider (submitted PR https://github.com/PowerShell/psl-omi-provider/pull/99, but since the CI build depends on powershell release package to run tests, the PR CI will fail until after beta.3 is out.)
- [ ] Remove `Microsoft.PowerShell.CoreCLR.AssemblyExtensions.LoadFrom(string assemblyPath)` once the `PackageManagement` module is migrated to .NET Core 2.0.
- [x] Move PowerShell Core to the latest .NET Core to get all fixes we need. (~via #3887~ Ran into a regression in the latest .NET Core, so moved back to 2.0.0-preview1 via #4026)
- [ ] Think about removing `Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.dll` completely.

### PR DESCRIPTION
**Pull request by daxian-dbw**
Fix #3649

Summary
----------
Move PowerShell to .NET Core `2.0.0-preview3-25426-01` by explicitly specify the `RuntimeFrameworkVersion` in .csproj files. `2.0.0-preview3` CLI doesn't work (see https://github.com/dotnet/cli/issues/7013, it's closed but I don't think it should be), so we use .NET CLI `2.0.0-preview2-006502` for now.

Also add `test/Test.Common.props` to hold common properties for test projects.

Note
-----
I have verified the CoreFx issues listed in #3649 are fixed in this .NET Core version.
I have verified that `PackageManagement` and `PowerShellGet` tests are all passed.
I have verified that `Login-AzureRmAccount` from `AzureRM.Netcore` works in powershell with this .NET Core version.

### CODE REVIEW

File: build.psm1 [hunk 1]
```diff
@@ -20,7 +20,7 @@ try {
 }
 
 $dotnetCLIChannel = "preview"
-$dotnetCLIRequiredVersion = "2.0.0-preview1-005952"
+$dotnetCLIRequiredVersion = "2.0.0-preview2-006502"
```
Comment (adityapatwardhan) [inline build.psm1]: preview3?

File: src/Microsoft.PowerShell.SDK/Microsoft.PowerShell.SDK.csproj [hunk 2]
```diff
@@ -17,17 +17,17 @@
...
-    <PackageReference Include="System.Net.Http.WinHttpHandler" Version="4.4.0-preview1-25302-01" />
-    <PackageReference Include="System.ServiceModel.Duplex" Version="4.4.0-preview1-25302-01" />
-    <PackageReference Include="System.ServiceModel.Http" Version="4.4.0-preview1-25302-01" />
-    <PackageReference Include="System.ServiceModel.NetTcp" Version="4.4.0-preview1-25302-01" />
-    <PackageReference Include="System.ServiceModel.Primitives" Version="4.4.0-preview1-25302-01" />
-    <PackageReference Include="System.ServiceModel.Security" Version="4.4.0-preview1-25302-01" />
-    <PackageReference Include="System.Text.Encodings.Web" Version="4.4.0-preview1-25302-01" />
-    <PackageReference Include="System.Threading.AccessControl" Version="4.4.0-preview1-25302-01" />
-    <PackageReference Include="System.Private.ServiceModel" Version="4.4.0-preview1-25302-01" />
+    <PackageReference Include="System.Data.SqlClient" Version="4.4.0-preview3-25423-02" />
```
Comment (adityapatwardhan) [inline src/Microsoft.PowerShell.SDK/Microsoft.PowerShell.SDK.csproj]: Is 02 expected? or should it be 01?

File: src/System.Management.Automation/System.Management.Automation.csproj [hunk 3]
```diff
@@ -11,12 +11,12 @@
   <ItemGroup>
     <ProjectReference Include="..\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext\Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.csproj" />
     <ProjectReference Include="..\Microsoft.PowerShell.CoreCLR.Eventing\Microsoft.PowerShell.CoreCLR.Eventing.csproj" />
-    <PackageReference Include="Microsoft.Win32.Registry.AccessControl" Version="4.4.0-preview1-25302-01" />
+    <PackageReference Include="Microsoft.Win32.Registry.AccessControl" Version="4.4.0-preview3-25423-02" />
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/System.Management.Automation.csproj]: Same here 02 or 01?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1 [hunk 4]
```diff
@@ -66,7 +66,9 @@ Describe "Get-ChildItem" -Tags "CI" {
             $file.Count | Should be 1
             $file.Name | Should be $item_F
         }
-        It "Should continue enumerating a directory when a contained item is deleted" {
+        # Test is pending on Unix platforms because of a behavior change in the latest .NET Core.
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1]: The issue #20456 is closed, do we still need to have the test case pending?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1 [hunk 5]
```diff
@@ -84,7 +86,9 @@ Describe "Get-ChildItem" -Tags "CI" {
                 $result.Count | Should BeExactly 4
             }
         }
-        It "Should continue enumerating a directory when a contained item is renamed" {
+        # Test is pending on Unix platforms because of a behavior change in the latest .NET Core.
+        # See https://github.com/dotnet/corefx/issues/20456 for more information.
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1]: Same as above.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: build.psm1 (hunk 1 above)
Comment (daxian-dbw) [inline build.psm1]: preview3 .NET CLI is still broken, please see the description of this PR.
Now we are using .NET CLI preivew2 to build, but target preview3 .NET Core Framework.

File: src/Microsoft.PowerShell.SDK/Microsoft.PowerShell.SDK.csproj (hunk 2 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.SDK/Microsoft.PowerShell.SDK.csproj]: Yes, it is expected. `Microsoft.NetCore.App 2.0.0-preview3-25426-01` references to packages of the build `25423-02` (built in 6/23). However, for `ServiceModel` related packages, they only have `25423-01` built on that day, so `25423-01` is used for `ServiceModel` related packages.

File: src/System.Management.Automation/System.Management.Automation.csproj (hunk 3 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/System.Management.Automation.csproj]: 02 is expected. Please see my response above.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1 (hunk 4 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1]: The behavior change was by design, to better align with Full .NET. So yes, this test and the corresponding fix need to be revisited. Here is the tracking issue: #4145

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1 (hunk 5 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-ChildItem.Tests.ps1]: Ditto.

**Comment by daxian-dbw** [comment]
@adityapatwardhan Your comments are addressed, please take another look when you have time. Thanks!

**Review by adityapatwardhan — APPROVED** [review]

**Comment by chuanjiao10** [comment]
Powershell 6.0 beta4 release notes, will it be added?

Because .net preview 2
Powershell support: (or experimental support)
• SUSE Linux Enterprise Server 12 SP2
• Debian 9
• Fedora 25 and 26 (beta)

I am looking forward

**Comment by daxian-dbw** [comment]
@chuanjiao10 new distro supports are tracked by #3961

**Comment by chuanjiao10** [comment]
daxian-dbw: ok 3q

**Pull request merged**